README.md
# countdown-timer README

[![Extension testing](https://github.com/ShanuDey/countdown-timer/actions/workflows/test.yml/badge.svg)](https://github.com/ShanuDey/countdown-timer/actions/workflows/test.yml)
[![Test and Deploy Extension](https://github.com/ShanuDey/countdown-timer/actions/workflows/deploy.yml/badge.svg)](https://github.com/ShanuDey/countdown-timer/actions/workflows/deploy.yml)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
[![VS Marketplace](https://vsmarketplacebadge.apphb.com/version/ShanuDey.countdown-timer.svg)](https://marketplace.visualstudio.com/items?itemName=ShanuDey.countdown-timer)
[![Installs](https://vsmarketplacebadge.apphb.com/installs/ShanuDey.countdown-timer.svg)](https://marketplace.visualstudio.com/items?itemName=ShanuDey.countdown-timer)

A vscode extension for better productivity and time management.

## Features

- Set timer for countdown and focus on your productivity
- Pop up a notification once countdown expirers. Take a break and come up with new ideas
- show and hide commands added for the status bar item
- Pomodoro timer included
- Customize pomodoro timer with custom work and break time

> Tip: This is a superb extension

## Requirements

It supports in all VS Code versions.

## Extension Settings

- `countdown-timer.activate`: To activate this extension manually (optional)
- `countdown-timer.settimer`: Set timer time for countdown
- `countdown-timer.hidestatusonidle`: Hide status bar item on idle
- `countdown-timer.showstatusonidle`: Show status bar item on idle
- `countdown-timer.startpomodoro`: Start Pomodoro mode
- `countdown-timer.stoppomodoro`: Stop Pomodoro mode
- `countdown-timer.update_pomodoro_work_time`: Update Pomodoro Work Time
- `countdown-timer.update_pomodoro_break_time`: Update Pomodoro Break Time

## Known Issues

There are no known issues

### For more information

- [Source Code](https://github.com/ShanuDey/countdown-timer)
- [Author](https://github.com/ShanuDey)

**Enjoy!**


CHANGELOG.md
# Change Log

## [1.0.0]

- Added Pomodoro timer
- Added commands to customize pomodoro timer work and break time
- Improved performance with new timer implementation logic

## [0.0.5]

- show and hide command added for status bar item
- handle visible state of status bar with the set timer

## [Unreleased]

- Initial release


test/suite/extension.test.js
const assert = require('assert');

// You can import and use all API from the 'vscode' module
// as well as import your extension to test it
const vscode = require('vscode');
const extension = require('../../extension.js');

suite('Countdown Timer Test Suite: validateTime function', () => {
  vscode.window.showInformationMessage('Start all tests.');

  test('test the input input following "xx:xx:xx" pattern', () => {
    // when the user input is not following this pattern
    assert.strictEqual(
      extension.validateTime('a:a:a'),
      'Invalid format!! Please enter time in HH:MM:SS format'
    );
    assert.strictEqual(
      extension.validateTime('aa:a:a'),
      'Invalid format!! Please enter time in HH:MM:SS format'
    );
    assert.strictEqual(
      extension.validateTime('aa:aa:a'),
      'Invalid format!! Please enter time in HH:MM:SS format'
    );
    assert.strictEqual(
      extension.validateTime('aa:aa;aa'),
      'Invalid format!! Please enter time in HH:MM:SS format'
    );
    // when the user input is following this pattern
    assert.strictEqual(extension.validateTime('11:11:11'), null);
  });

  test('test the user inputs are digit', () => {
    // when the user inputs are not digits
    assert.strictEqual(
      extension.validateTime('aa:aa:aa'),
      'Invalid format!! Please enter time in HH:MM:SS format'
    );
    // when the user inputs are digits
    assert.strictEqual(extension.validateTime('11:11:11'), null);
  });

  test('test the user input is within hours, minutes and seconds max range limitation', () => {
    assert.strictEqual(
      extension.validateTime('11:11:60'),
      'Seconds are more than 59!! Add Minutes instead'
    );
    assert.strictEqual(
      extension.validateTime('10:60:59'),
      'Minutes are more than 59!! Add Hours instead'
    );
    assert.strictEqual(
      extension.validateTime('25:00:00'),
      'Hours are more than 24!! More than a day is not supported yet.'
    );
    assert.strictEqual(
      extension.validateTime('24:00:01'),
      'Hours are more than 24!! More than a day is not supported yet.'
    );
    assert.strictEqual(
      extension.validateTime('24:01:00'),
      'Hours are more than 24!! More than a day is not supported yet.'
    );
    assert.strictEqual(
      extension.validateTime('00:00:00'),
      'OMG!! All zeros. It is already expired'
    );
    assert.strictEqual(extension.validateTime('24:00:00'), null);
  });
});


extension.js
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
const { clearInterval } = require('timers');
const vscode = require('vscode');

let statusBar;
const commandId = 'countdown-timer.activate';
const setTimerCommandId = 'countdown-timer.settimer';
const hideStatusBarOnIdleCommandId = 'countdown-timer.hidestatusonidle';
const showStatusBarOnIdleCommandId = 'countdown-timer.showstatusonidle';
const startPomodoroTimerCommandId = 'countdown-timer.startpomodoro';
const stopPomodoroTimerCommandId = 'countdown-timer.stoppomodoro';
const updatePomodoroWorkTimeCommandId =
  'countdown-timer.update_pomodoro_work_time';
const updatePomodoroBreakTimeCommandId =
  'countdown-timer.update_pomodoro_break_time';

let pomodoroTimerIntervalId;
let isBreak = false;
let WorkTime = 1500; // 25 min
let BreakTime = 300; // 5 min
let timeLeft = 0;
let timerIntervalId;

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed

/**
 * @param {vscode.ExtensionContext} context
 */
function activate(context) {
  // The command has been defined in the package.json file
  // Now provide the implementation of the command with  registerCommand
  // The commandId parameter must match the command field in package.json
  context.subscriptions.push(disposable);
  context.subscriptions.push(setTimerRegisterCommand);
  // create a new status bar item that we can now manage
  context.subscriptions.push(statusBar);
  context.subscriptions.push(hideStatusBarItemOnIdleCommand);
  context.subscriptions.push(showStatusBarItemOnIdleCommand);
  context.subscriptions.push(startPomodoroTimerCommand);
  context.subscriptions.push(stopPomodoroTimerCommand);
  context.subscriptions.push(updatePomodoroWorkTimeCommand);
  context.subscriptions.push(updatePomodoroBreakTimeCommand);

  //on active
  createStatusBar();
}

// this method is called when your extension is deactivated
function deactivate() {}

let disposable = vscode.commands.registerCommand(commandId, function () {
  // Display a message box to the user
  vscode.window.showInformationMessage(
    'Count Down timer extension is active now!'
  );
});

const setTimerRegisterCommand = vscode.commands.registerCommand(
  setTimerCommandId,
  async function () {
    const userInput = await getUserInput(
      'HH:MM:SS in 24 hours format. Example: 00:30:00',
      validateTime
    );
    vscode.window.showInformationMessage(
      `Countdown timer is set for ${userInput}  ⬇️`
    );
    manageTimer(userInput);
  }
);

const hideStatusBarItemOnIdleCommand = vscode.commands.registerCommand(
  hideStatusBarOnIdleCommandId,
  async function () {
    statusBar.hide();
  }
);

const showStatusBarItemOnIdleCommand = vscode.commands.registerCommand(
  showStatusBarOnIdleCommandId,
  function () {
    statusBar.show();
  }
);

const startPomodoroTimerCommand = vscode.commands.registerCommand(
  startPomodoroTimerCommandId,
  function () {
    pomodoroTimer();
  }
);

const stopPomodoroTimerCommand = vscode.commands.registerCommand(
  stopPomodoroTimerCommandId,
  function () {
    clearInterval(pomodoroTimerIntervalId);
    updateStatusBar('Not Set');
  }
);

const updatePomodoroWorkTimeCommand = vscode.commands.registerCommand(
  updatePomodoroWorkTimeCommandId,
  async function () {
    const userInput = await getUserInput(
      'HH:MM:SS in 24 hours format. Example: 00:30:00',
      validateTime
    );
    WorkTime = getTimeInSeconds(userInput);
    console.log(WorkTime);
    restartPomodoroTimer();
  }
);

const updatePomodoroBreakTimeCommand = vscode.commands.registerCommand(
  updatePomodoroBreakTimeCommandId,
  async function () {
    const userInput = await getUserInput(
      'HH:MM:SS in 24 hours format. Example: 00:30:00',
      validateTime
    );
    BreakTime = getTimeInSeconds(userInput);
    console.log(BreakTime);
    restartPomodoroTimer();
  }
);

function createStatusBar() {
  statusBar = vscode.window.createStatusBarItem(
    vscode.StatusBarAlignment.Right
  );
  updateStatusBar('Not set');
  statusBar.show();
  statusBar.command = setTimerCommandId;
}

function updateStatusBar(text) {
  statusBar.text = `$(watch) ${text}`;
  // console.log(statusBar.text);
}

const validateTime = (inputTime) => {
  // checking the pattern is `dd:dd:dd` where d is digit
  const pattern = /\d\d:\d\d:\d\d/g;
  if (!pattern.test(inputTime))
    return 'Invalid format!! Please enter time in HH:MM:SS format';

  // checking part wise
  const [hh, mm, ss] = inputTime.split(':').map((str) => parseInt(str));
  if (ss > 59) return 'Seconds are more than 59!! Add Minutes instead';
  if (mm > 59) return 'Minutes are more than 59!! Add Hours instead';
  if (hh > 24)
    return 'Hours are more than 24!! More than a day is not supported yet.';
  if (hh === 24 && mm >= 0 && ss > 0)
    return 'Hours are more than 24!! More than a day is not supported yet.';
  if (hh === 24 && mm > 0 && ss >= 0)
    return 'Hours are more than 24!! More than a day is not supported yet.';
  if (hh === 0 && mm === 0 && ss === 0)
    return 'OMG!! All zeros. It is already expired';

  // if all test passed then return null
  return null;
};

/**
 * Shows an input box to take user input
 */
async function getUserInput(placeHolderText, validateInputFunction) {
  const result = await vscode.window.showInputBox({
    // value: 'abcdef',
    // valueSelection: [2, 4],
    placeHolder: placeHolderText,
    validateInput: (text) => {
      return validateInputFunction(text);
    },
  });
  return result;
}

function manageTimer(targetTime) {
  const statusbarVisibility = statusBar._visible;
  if (statusbarVisibility === false) statusBar.show();
  let currentTimeLeft;
  timeLeft = getTimeInSeconds(targetTime);
  timerIntervalId = setInterval(() => {
    if (timeLeft < 0) {
      clearInterval(timerIntervalId);
      updateStatusBar('Not Set');
      vscode.window.showInformationMessage(
        'Countdown Timer: Expired \n 🎉🎉 Congratulation 🎉🎉'
      );
      statusbarVisibility ? statusBar.show() : statusBar.hide();
    } else {
      currentTimeLeft = getCurrentTimeLeft();
      updateStatusBar(currentTimeLeft);
      timeLeft--;
    }
  }, 1000);
}

const getCurrentTimeLeft = () => {
  const hh = Math.floor(timeLeft / 3600);
  const mm = Math.floor(timeLeft / 60);
  const ss = Math.floor(timeLeft % 60);
  return `${hh}h ${mm}m ${ss}s`;
};

const getTimeInSeconds = (targetTime) => {
  const [hh, mm, ss] = targetTime.split(':').map((str) => parseInt(str));
  return hh * 60 + mm * 60 + ss;
};

const pomodoroTimer = () => {
  timeLeft = WorkTime;
  pomodoroTimerIntervalId = setInterval(() => {
    if (timeLeft < 0) {
      if (isBreak) {
        timeLeft = WorkTime;
        vscode.window.showInformationMessage(
          'Countdown Timer: Focusing in work'
        );
      } else {
        timeLeft = BreakTime;
        vscode.window.showInformationMessage('Countdown Timer: Have a break');
      }
      isBreak = !isBreak;
    }
    updateStatusBar(getCurrentTimeLeft());
    timeLeft--;
  }, 1000);
};

const restartPomodoroTimer = () => {
  clearInterval(pomodoroTimerIntervalId);
  updateStatusBar(`Restarting $(debug-restart)`);
  pomodoroTimer();
};

module.exports = {
  activate,
  deactivate,
  validateTime,
};
